### PR TITLE
Implementation of loading UGC data for generator

### PR DESCRIPTION
**Pull request by syershov**

### CODE REVIEW

File: generator/generator_tests/generator_tests.pro [hunk 1]
```diff
@@ -47,3 +47,5 @@ SOURCES += \
     tag_admixer_test.cpp \
     tesselator_test.cpp \
     triangles_tree_coding_test.cpp \
+	ugc_test.cpp \
```
Comment (ygorshenin) [inline generator/generator_tests/generator_tests.pro]: Format, please.

File: generator/generator_tests/ugc_test.cpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
+#include "testing/testing.hpp"
+
+#include "generator/ugc_db.hpp"
+#include "generator/ugc_translator.hpp"
+#include "generator/osm_id.hpp"
+
+#include "ugc/types.hpp"
+
+
```
Comment (ygorshenin) [inline generator/generator_tests/ugc_test.cpp]: Remove this blank line, please.

File: generator/generator_tests/ugc_test.cpp [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+std::string database(R"LLL(
```
Comment (ygorshenin) [inline generator/generator_tests/ugc_test.cpp]: g_database

File: generator/ugc_db.cpp [hunk 4]
```diff
@@ -1,15 +1,98 @@
 #include "generator/ugc_db.hpp"
 
+#include "base/logging.hpp"
 #include "base/macros.hpp"
 
+#include <iostream>
```
Comment (ygorshenin) [inline generator/ugc_db.cpp]: You need sstream here.

File: generator/ugc_db.cpp [hunk 5]
```diff
@@ -1,15 +1,98 @@
...
+
+namespace {
```
Comment (ygorshenin) [inline generator/ugc_db.cpp]: clang-format please.

File: generator/ugc_db.cpp [hunk 6]
```diff
@@ -1,15 +1,98 @@
...
+}
+
+UGCDB::UGCDB(std::string const & path)
+{
+  int rc;
+
+  rc = sqlite3_open(path.c_str(), &m_db);
+  if(rc)
+  {
+    LOG(LERROR, ("Can't open database:", sqlite3_errmsg(m_db)));
```
Comment (ygorshenin) [inline generator/ugc_db.cpp]: How about exceptions here?

Comment (ygorshenin) [inline generator/ugc_db.cpp]: At least, create a helpere method named "IsOpen" in the UGCDB.

File: generator/ugc_db.hpp [hunk 7]
```diff
@@ -20,6 +24,12 @@ class UGCDB
 {
 public:
   UGCDB(std::string const & path);
+  ~UGCDB();
   WARN_UNUSED_RESULT bool Get(osm::Id const & id, std::vector<uint8_t> & blob);
+  WARN_UNUSED_RESULT bool Exec(std::string const & statement);
+private:
```
Comment (ygorshenin) [inline generator/ugc_db.hpp]: Blank line before private.

File: generator/ugc_db.hpp [hunk 8]
```diff
@@ -20,6 +24,12 @@ class UGCDB
...
+  bool ValueToBlob(std::string const & src, std::vector<uint8_t> & blob);
+private:
```
Comment (ygorshenin) [inline generator/ugc_db.hpp]: No need in second private section.

File: generator/ugc_db.cpp (hunk 5 above)
Comment (syershov) [inline generator/ugc_db.cpp]: done

File: generator/ugc_db.hpp [hunk 9]
```diff
@@ -20,6 +24,12 @@ class UGCDB
...
+  sqlite3 * m_db = nullptr;
```
Comment (ygorshenin) [inline generator/ugc_db.hpp]: Just forward-declare sqlite3 and move inclusion of sqlite.h to the cpp.

File: generator/ugc_db.cpp (hunk 4 above)
Comment (syershov) [inline generator/ugc_db.cpp]: done

File: generator/generator_tests/ugc_test.cpp (hunk 3 above)
Comment (syershov) [inline generator/generator_tests/ugc_test.cpp]: done

File: generator/generator_tests/ugc_test.cpp (hunk 2 above)
Comment (syershov) [inline generator/generator_tests/ugc_test.cpp]: done

File: generator/ugc_translator.cpp [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+  for (size_t i=0; i<size; ++i)
+  {
+    json_t * el = json_array_get(jsonRoot.get(), i);
+    double ratingValue = 0;
+    size_t translationKeyId = 0;
+
+    FromJSONObject(el, "value", ratingValue);
+    FromJSONObject(el, "criterion_id", translationKeyId);
+
+    std::stringstream translationKey;
```
Comment (ygorshenin) [inline generator/ugc_translator.cpp]: ostringstream please.

File: generator/ugc_translator.cpp [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+{
+  return true;
+}
+
+//bool UGCTranslator::TranslateAttribute(UGCDB & db, osm::Id const id, ugc::Attribute & attribute)
+//{
+//  return false;
+//}
+}  // namespace generator
+
```
Comment (ygorshenin) [inline generator/ugc_translator.cpp]: No need in this blank line.

File: generator/ugc_db.cpp (hunk 6 above)
Comment (syershov) [inline generator/ugc_db.cpp]: I think, no.

File: generator/ugc_db.hpp (hunk 7 above)
Comment (syershov) [inline generator/ugc_db.hpp]: done

File: generator/ugc_db.cpp [hunk 12]
```diff
@@ -1,15 +1,98 @@
...
+}
+
+bool UGCDB::Exec(std::string const & statement)
+{
+  if(!m_db)
+    return false;
+
+  char *zErrMsg = 0;
+  auto rc = sqlite3_exec(m_db, statement.c_str(), nullptr, nullptr, &zErrMsg);
+  if( rc!=SQLITE_OK ){
```
Comment (ygorshenin) [inline generator/ugc_db.cpp]: format please.

File: generator/ugc_db.cpp [hunk 13]
```diff
@@ -1,15 +1,98 @@
...
+    return false;
+  }
+  results.values << "]";
+
+  return ValueToBlob(results.values.str(), blob);
+}
+
+bool UGCDB::Exec(std::string const & statement)
+{
+  if(!m_db)
```
Comment (ygorshenin) [inline generator/ugc_db.cpp]: format please.

File: generator/ugc_db.hpp (hunk 8 above)
Comment (syershov) [inline generator/ugc_db.hpp]: done

File: generator/ugc_db.hpp (hunk 9 above)
Comment (syershov) [inline generator/ugc_db.hpp]: done

File: generator/ugc_translator.cpp (hunk 10 above)
Comment (syershov) [inline generator/ugc_translator.cpp]: done

File: generator/ugc_translator.cpp (hunk 11 above)
Comment (syershov) [inline generator/ugc_translator.cpp]: done

File: generator/ugc_db.cpp (hunk 12 above)
Comment (syershov) [inline generator/ugc_db.cpp]: done

File: generator/ugc_db.cpp (hunk 13 above)
Comment (syershov) [inline generator/ugc_db.cpp]: done

File: generator/generator_tests/generator_tests.pro [hunk 14]
```diff
@@ -47,3 +47,5 @@ SOURCES += \
     tag_admixer_test.cpp \
     tesselator_test.cpp \
     triangles_tree_coding_test.cpp \
+    ugc_test.cpp \
+
```
Comment (ygorshenin) [inline generator/generator_tests/generator_tests.pro]: Remove trailing newline please.

File: generator/ugc_db.cpp [hunk 15]
```diff
@@ -1,15 +1,101 @@
 #include "generator/ugc_db.hpp"
 
+#include "base/logging.hpp"
 #include "base/macros.hpp"
 
+#include <sqlite3.h>
+
+#include <iostream>
```
Comment (ygorshenin) [inline generator/ugc_db.cpp]: You don't need iostream. Also, please, replace stringstreams by ostringstreams.

File: generator/ugc_db.cpp [hunk 16]
```diff
@@ -1,15 +1,101 @@
...
+#include <sstream>
+
+namespace
+{
+struct Results
+{
+  size_t counter = 0;
```
Comment (ygorshenin) [inline generator/ugc_db.cpp]: Please, replace counter by "bool m_empty"

File: generator/ugc_db.hpp [hunk 17]
```diff
@@ -20,6 +24,13 @@ class UGCDB
 {
 public:
   UGCDB(std::string const & path);
+  ~UGCDB();
   WARN_UNUSED_RESULT bool Get(osm::Id const & id, std::vector<uint8_t> & blob);
```
Comment (ygorshenin) [inline generator/ugc_db.hpp]: Blank line before method Get, please.

File: generator/ugc_translator.cpp [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+  std::vector<uint8_t> blob;
+  bool rc = db.Get(id, blob);
+  if (!rc)
+    return false;
+  std::string result(blob.cbegin(), blob.cend());
+
+  my::Json jsonRoot(result);
+
+  size_t size = json_array_size(jsonRoot.get());
+  for (size_t i=0; i<size; ++i)
```
Comment (ygorshenin) [inline generator/ugc_translator.cpp]: format please.

File: generator/ugc_translator.hpp [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+
+  bool TranslateUGC(osm::Id const & id, ugc::UGC & ugc);
+  // For testing only
+  void CreateRatings(std::string const & data);
+  void CreateReviews(std::string const & data);
+private:
+  bool TranslateRating(UGCDB & db, osm::Id const id, ugc::Rating & rating);
+  bool TranslateReview(UGCDB & db, osm::Id const id, std::vector<ugc::Review> & review);
+//  bool TranslateAttribute(UGCDB & db, osm::Id const id, ugc::Attribute & attribute);
+private:
```
Comment (ygorshenin) [inline generator/ugc_translator.hpp]: Blank line before private please.

File: generator/generator_tests/generator_tests.pro (hunk 14 above)
Comment (syershov) [inline generator/generator_tests/generator_tests.pro]: done

File: generator/ugc_db.cpp (hunk 16 above)
Comment (syershov) [inline generator/ugc_db.cpp]: done

File: generator/ugc_db.hpp (hunk 17 above)
Comment (syershov) [inline generator/ugc_db.hpp]: done

File: generator/ugc_translator.cpp (hunk 18 above)
Comment (syershov) [inline generator/ugc_translator.cpp]: done

File: generator/ugc_translator.hpp (hunk 19 above)
Comment (syershov) [inline generator/ugc_translator.hpp]: done

**Comment by syershov** [comment]
@ygorshenin PTAL

File: generator/ugc_db.cpp [hunk 20]
```diff
@@ -1,15 +1,100 @@
...
+  }
+  return 0;
 }
 
-bool UGCDB::Get(osm::Id const & /* id */, std::vector<uint8_t> & /* blob */)
-{  // TODO (@syershov): implement this
-  return false;
+UGCDB::UGCDB(std::string const & path)
+{
+  int rc;
```
Comment (mgsergio) [inline generator/ugc_db.cpp]: auto/int rc = sqlite3_open(path.c_str(), &m_db);

File: generator/ugc_db.cpp [hunk 21]
```diff
@@ -1,15 +1,100 @@
...
+};
+}  // namespace
+
 namespace generator
 {
-UGCDB::UGCDB(std::string const & /*path */)
-{  // TODO (@syershov): implement this
+static int callback(void * results_ptr, int argc, char ** argv, char ** azColName)
+{
+  Results & results = *reinterpret_cast<Results *>(results_ptr);
```
Comment (mgsergio) [inline generator/ugc_db.cpp]: auto &

File: generator/ugc_db.cpp [hunk 22]
```diff
@@ -1,15 +1,100 @@
...
+{
+  int rc;
+
+  rc = sqlite3_open(path.c_str(), &m_db);
+  if (rc)
+  {
+    LOG(LERROR, ("Can't open database:", sqlite3_errmsg(m_db)));
+    sqlite3_close(m_db);
+    m_db = nullptr;
+    return;
```
Comment (mgsergio) [inline generator/ugc_db.cpp]: No need in return here

File: generator/ugc_db.cpp [hunk 23]
```diff
@@ -1,15 +1,100 @@
...
+  if (m_db)
+    sqlite3_close(m_db);
 }
+
+bool UGCDB::Get(osm::Id const & id, std::vector<uint8_t> & blob)
+{
+  if (!m_db)
+    return false;
+
+  char * zErrMsg = 0;
```
Comment (mgsergio) [inline generator/ugc_db.cpp]: nullptr

File: generator/ugc_db.cpp (hunk 20 above)
Comment (syershov) [inline generator/ugc_db.cpp]: done

File: generator/ugc_db.cpp (hunk 21 above)
Comment (syershov) [inline generator/ugc_db.cpp]: skipped

File: generator/ugc_db.cpp (hunk 22 above)
Comment (syershov) [inline generator/ugc_db.cpp]: done

File: generator/ugc_db.cpp (hunk 23 above)
Comment (syershov) [inline generator/ugc_db.cpp]: done

**Comment by syershov** [comment]
@mgsergio PTAL

**Pull request merged**